### PR TITLE
feat: add abac_with_rule and test

### PR DESCRIPTION
**Pull request by sheny1xuan**
Fix: https://github.com/casbin/casbin-cpp/issues/102

### Description
+ feat: add abac_with_rule and test

### CODE REVIEW

**Comment by hsluoyz** [comment]
@sheny1xuan what issue does this PR fix?

**Comment by sheny1xuan** [comment]
This PR is implement of ABAC `eavl()` key word, it's continue work of  #102 .

File: casbin/util/eval.cpp [hunk 1]
```diff
@@ -60,6 +60,25 @@ std::string ReplaceEvalWithMap(const std::string& src, std::unordered_map<std::s
     return replacedExp;
 }
 
+// GetEvalValue returns the parameters of function eval
+std::vector<std::string> GetEvalValue(const std::string& s) {
+    std::vector<std::string> rules;
```
Comment (EmperorYP7) [inline casbin/util/eval.cpp:65]:
```suggestion
    std::vector<std::string> rules;
    rules.reserve(100);  // 100 or any arbitrary number
```

File: casbin/util/eval.cpp [hunk 2]
```diff
@@ -60,6 +60,25 @@ std::string ReplaceEvalWithMap(const std::string& src, std::unordered_map<std::s
...
+    std::string srcCpy = s;
```
Comment (EmperorYP7) [inline casbin/util/eval.cpp:66]: instead, take the param as a copy.

```cpp
std::vector<std::string> GetEvalValue(std::string s)
```

File: casbin/util/util.h [hunk 3]
```diff
@@ -71,6 +71,8 @@ bool HasEval(const std::string& s);
 // ReplaceEvalWithMap replace function eval with the value of its parameters via given sets.
 std::string ReplaceEvalWithMap(const std::string& src, std::unordered_map<std::string, std::string>& sets);
 
+// GetEvalValue returns the parameters of function eval
+std::vector<std::string> GetEvalValue(const std::string& s);
```
Comment (EmperorYP7) [inline casbin/util/util.h:75]:
```suggestion
std::vector<std::string> GetEvalValue(std::string s); 
```

File: include/casbin/casbin_helpers.h [hunk 4]
```diff
@@ -640,6 +640,9 @@ namespace casbin {
 
     // ReplaceEvalWithMap replace function eval with the value of its parameters via given sets.
     std::string ReplaceEvalWithMap(const std::string& src, std::unordered_map<std::string, std::string>& sets);
+    
+    // GetEvalValue returns the parameters of function eval
+    std::vector<std::string> GetEvalValue(const std::string& s);
```
Comment (EmperorYP7) [inline include/casbin/casbin_helpers.h:645]: likewise

File: tests/config_path.h [hunk 5]
```diff
@@ -53,3 +53,5 @@ static const std::string priority_model_path = relative_path + "/examples/priori
 static const std::string priority_policy_path = relative_path + "/examples/priority_policy.csv";
 
 static const std::string abac_model_path = relative_path + "/examples/abac_model.conf";
+static const std::string abac_rule_model_path = relative_path + "/examples/abac_rule_model.conf";
+static const std::string abac_rule_policy_path = relative_path + "/examples/abac_rule_policy.csv";
```
Comment (EmperorYP7) [inline tests/config_path.h:57]:
```suggestion
static const std::string abac_rule_policy_path = relative_path + "/examples/abac_rule_policy.csv";

```

**Review by EmperorYP7 — CHANGES_REQUESTED** [review]
Great! Just a few minor tweaks, and its ready to go.

**Comment by EmperorYP7** [comment]
I think #102 is on the verge of completion. cc @carrionbent

**Comment by hsluoyz** [comment]
@carrionbent plz review

File: casbin/util/eval.cpp (hunk 1 above)
Comment (sheny1xuan) [reply casbin/util/eval.cpp:65]: Maybe this only needs only a few capacity for `eval()`, `10` is ok I think.

File: casbin/util/eval.cpp (hunk 2 above)
Comment (sheny1xuan) [reply casbin/util/eval.cpp]: Fixed it.

**Review by EmperorYP7 — APPROVED** [review]
LGTM

**Comment by hsluoyz** [comment]
It seems that @carrionbent is no longer active for half a year. We can merge this first.

**Pull request merged**